framework/src/org/checkerframework/common/fieldtypeinference/FieldTypeInferenceTransfer.java

```diff
@@ -0,0 +1,112 @@
+ * <tt>org.checkerframework.common.fieldtypeinference.FieldTypeInferenceAnnotatedTypeFactory</tt>.
+ *
+ * @author pbsf
+ */
+
+public class FieldTypeInferenceTransfer extends CFTransfer {
+
+    private final Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields;
+    private final FieldTypeInferenceAnnotatedTypeFactory factory;
+
+    public FieldTypeInferenceTransfer(CFAbstractAnalysis<CFValue, CFStore,
+            CFTransfer> analysis) {
+        super(analysis);
+        assert (analysis.getTypeFactory() instanceof FieldTypeInferenceAnnotatedTypeFactory);
+        factory = (FieldTypeInferenceAnnotatedTypeFactory) analysis.getTypeFactory();
+        assignedPrivateFields = factory.getAssignedPrivateFields();
+    }
+
+    @Override
+    public CFStore initialStore(UnderlyingAST underlyingAST,
+            List<LocalVariableNode> parameters) {
+        CFStore initialStore = super.initialStore(underlyingAST, parameters);
+        for (Receiver r : assignedPrivateFields.keySet()) {
+            for (AnnotationMirror am : assignedPrivateFields.get(r).
+                    getAnnotations()) {
+                initialStore.insertValue(r, am);
+            }
+        }
+        return initialStore;
+    }
+
+    @Override
+    public TransferResult<CFValue, CFStore> visitAssignment(AssignmentNode n,
+            TransferInput<CFValue, CFStore> in) {
+        TransferResult<CFValue, CFStore> result = super.visitAssignment(n, in);
+        assert result instanceof RegularTransferResult;
+        Node rhs = n.getExpression();
+        Receiver expr = FlowExpressions.internalReprOf(analysis.getTypeFactory(),
+                n.getTarget());
+
+        // If this is an assignment to a field of 'this' and this field is
+        // private, then populate the map in the ATF.
+        if (!expr.containsUnknown() && expr instanceof FieldAccess &&
+                ((FieldAccess) expr).getField().getModifiers().
+                contains(Modifier.PRIVATE)) {
+            addAssignedField(expr, factory.getAnnotatedType(rhs.getTree()));
+        }
+        return result;
+    }
+
+    /**
+     * Update the ATM mapped by the parameter <tt>r</tt> in the map
+     * <tt>ClassValAnnotatedTypeFactory.analysis.assignedPrivateFields</tt> to be the LUB
+     * of the parameter <tt>atm</tt> and the current ATM mapped by <tt>r</tt>
+     * in this map.
+     * <p>
+     * If <tt>r</tt> is not a key in the map, it is added to the map
+     * mapping to <tt>atm</tt>.
+     * @param r
+     *      A receiver representing a private field.
+     * @param atm
+     *      An ATM assigned to the private field in <tt>r</tt>.
+     */
+    public void addAssignedField(Receiver r, AnnotatedTypeMirror atm) {
+        Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields = factory.getAssignedPrivateFields();
+        if (assignedPrivateFields.containsKey(r)) {
+            Set<? extends AnnotationMirror> lub = analysis.getTypeFactory().
```

Comment:
This comment makes a lot of sense, and I agree with everything. But unlike the other comments, I'm not sure how to fix this one. I understand that we must take the lowerBounds and upperBounds of the AnnotatedTypeMirror in consideration, as well as the primary annotations, but how to do that correctly using the Checker Framework? I don't see a LUB method for AnnotatedTypeMirror in the QualifierHierarchy. The closest one I found is this: 
```
public Set<? extends AnnotationMirror> leastUpperBounds(AnnotatedTypeMirror type1,
            AnnotatedTypeMirror type2, Collection<? extends AnnotationMirror> annos1,
            Collection<AnnotationMirror> annos2)
```
But as you can see, only primary annotations are returned. 
Is there a reason why a method that calculates the LUB between two AnnotatedTypeMirror and returns an AnnotatedTypeMirror does not exist? In case not, should we implement it?

Or does a possible solution to that problem has anything to do with the AnnotatedTypeMerger you mentioned in another comment? (The comment about using the AnnotatedTypeMerger wasn't very clear to me to be honest.)

Thanks a lot for all the feedback!